commitrmnocb/types.go

```diff
@@ -0,0 +1,255 @@
+		}
+
+		// FChain
+		for chainSel, f := range obs.FChain {
+			aggObs.FChain[chainSel] = append(aggObs.FChain[chainSel], f)
+		}
+	}
+
+	return aggObs
+}
+
+// ConsensusObservation holds the consensus values for all chains across all observations in a round
+type ConsensusObservation struct {
+	// A map from chain selectors to each chain's consensus merkle root
+	MerkleRoots map[cciptypes.ChainSelector]cciptypes.MerkleRootChain
+
+	// A map from chain selectors to each chain's consensus gas prices
+	GasPrices map[cciptypes.ChainSelector]cciptypes.BigInt
+
+	// A map from token IDs to each token's consensus price
+	TokenPrices map[types.Account]cciptypes.BigInt
+
+	// A map from chain selectors to each chain's consensus OnRamp max sequence number
+	OnRampMaxSeqNums map[cciptypes.ChainSelector]cciptypes.SeqNum
+
+	// A map from chain selectors to each chain's consensus OffRamp next sequence number
+	OffRampNextSeqNums map[cciptypes.ChainSelector]cciptypes.SeqNum
+
+	// A map from chain selectors to each chain's consensus f (failure tolerance)
+	FChain map[cciptypes.ChainSelector]int
+}
+
+// GasPricesSortedArray returns a sorted list of gas prices
+func (co ConsensusObservation) GasPricesSortedArray() []cciptypes.GasPriceChain {
+	gasPrices := make([]cciptypes.GasPriceChain, 0, len(co.GasPrices))
+	for chain, gasPrice := range co.GasPrices {
+		gasPrices = append(gasPrices, cciptypes.NewGasPriceChain(gasPrice.Int, chain))
+	}
+
+	sort.Slice(gasPrices, func(i, j int) bool {
+		return gasPrices[i].ChainSel < gasPrices[j].ChainSel
+	})
+
+	return gasPrices
+}
+
+// TokenPricesSortedArray returns a sorted list of token prices
+func (co ConsensusObservation) TokenPricesSortedArray() []cciptypes.TokenPrice {
+	tokenPrices := make([]cciptypes.TokenPrice, 0, len(co.TokenPrices))
+	for tokenID, tokenPrice := range co.TokenPrices {
+		tokenPrices = append(tokenPrices, cciptypes.NewTokenPrice(tokenID, tokenPrice.Int))
+	}
+
+	sort.Slice(tokenPrices, func(i, j int) bool {
+		return tokenPrices[i].TokenID < tokenPrices[j].TokenID
+	})
+
+	return tokenPrices
+}
+
+type OutcomeType int
+
+const (
+	ReportIntervalsSelected OutcomeType = iota
+	ReportGenerated
+	ReportEmpty
+	ReportInFlight
+	ReportTransmitted
+	ReportTransmissionFailed
+)
+
+type Outcome struct {
+	OutcomeType                     OutcomeType                 `json:"outcomeType"`
+	RangesSelectedForReport         []ChainRange                `json:"rangesSelectedForReport"`
+	RootsToReport                   []cciptypes.MerkleRootChain `json:"rootsToReport"`
+	OffRampNextSeqNums              []plugintypes.SeqNumChain   `json:"offRampNextSeqNums"`
+	TokenPrices                     []cciptypes.TokenPrice      `json:"tokenPrices"`
+	GasPrices                       []cciptypes.GasPriceChain   `json:"gasPrices"`
+	ReportTransmissionCheckAttempts uint                        `json:"reportTransmissionCheckAttempts"`
+}
+
+// Encode TODO: sort all lists here to ensure deterministic serialization
```

Comment:
I think we should do this in this PR so we don't run into nondeterminism issues down the line